File: app/interlink/repository.py
"""Data access for the interlink module (PostgreSQL via SQLAlchemy)."""

from __future__ import annotations

import uuid
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import datetime
from typing import Protocol

from sqlalchemy import func, select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session, undefer

from app.core.exceptions import ConflictError
from app.interlink.models import ACTIVE_STATUSES, InternalLinkSuggestion, SuggestionStatus
from app.pages.models import Page


@dataclass(frozen=True)
class SuggestionFilters:
    status: SuggestionStatus | None = None
    site_id: uuid.UUID | None = None
    source_page_id: uuid.UUID | None = None
    target_page_id: uuid.UUID | None = None
    min_relevance_score: int | None = None


class InterlinkRepository(Protocol):
    def get_page(self, page_id: uuid.UUID, *, for_update: bool = False) -> Page | None: ...

    def list_candidate_pool(self, source: Page) -> list[Page]: ...

    def load_contents(self, page_ids: Sequence[uuid.UUID]) -> dict[uuid.UUID, str]: ...

    def get_suggestion(
        self, suggestion_id: uuid.UUID, *, for_update: bool = False
    ) -> InternalLinkSuggestion | None: ...

    def list_suggestions(
        self, filters: SuggestionFilters, *, offset: int, limit: int
    ) -> tuple[list[InternalLinkSuggestion], int]: ...

    def active_target_ids(self, source_page_id: uuid.UUID) -> set[uuid.UUID]: ...

    def rejected_target_ids_since(
        self, source_page_id: uuid.UUID, since: datetime
    ) -> set[uuid.UUID]: ...

    def anchors_by_target(self, target_ids: Sequence[uuid.UUID]) -> dict[uuid.UUID, list[str]]: ...

    def add_suggestion(self, suggestion: InternalLinkSuggestion) -> bool:
        """Insert; return False (without failing the transaction) on an active-pair duplicate."""
        ...

    def commit(self) -> None:
        """Commit; raise ConflictError(ACTIVE_SUGGESTION_EXISTS) on an active-pair clash."""
        ...

    def rollback(self) -> None: ...


ACTIVE_PAIR_INDEX = "uq_internal_link_suggestions_active_pair"


def _constraint_name(exc: IntegrityError) -> str | None:
    diag = getattr(exc.orig, "diag", None)
    return getattr(diag, "constraint_name", None)


class SqlAlchemyInterlinkRepository:
    def __init__(self, session: Session) -> None:
        self._session = session

    def get_page(self, page_id: uuid.UUID, *, for_update: bool = False) -> Page | None:
        stmt = select(Page).where(Page.id == page_id).options(undefer(Page.content_html))
        if for_update:
            stmt = stmt.with_for_update()
        return self._session.scalar(stmt)

    def list_candidate_pool(self, source: Page) -> list[Page]:
        # Cheap SQL pre-filter; `candidate_filter` remains the authoritative rule set.
        stmt = (
            select(Page)
            .where(
                Page.site_id == source.site_id,
                Page.id != source.id,
                Page.http_status >= 200,
                Page.http_status < 300,
                Page.redirect_url.is_(None),
                Page.has_noindex.is_(False),
                Page.is_indexable.is_(True),
            )
            .order_by(Page.url)
        )
        return list(self._session.scalars(stmt))

    def load_contents(self, page_ids: Sequence[uuid.UUID]) -> dict[uuid.UUID, str]:
        if not page_ids:
            return {}
        rows = self._session.execute(
            select(Page.id, Page.content_html).where(Page.id.in_(page_ids))
        )
        return {row.id: row.content_html for row in rows if row.content_html}

    def get_suggestion(
        self, suggestion_id: uuid.UUID, *, for_update: bool = False
    ) -> InternalLinkSuggestion | None:
        stmt = select(InternalLinkSuggestion).where(InternalLinkSuggestion.id == suggestion_id)
        if for_update:
            # Lock only the suggestion row, not the joined page rows.
            stmt = stmt.with_for_update(of=InternalLinkSuggestion)
        return self._session.scalar(stmt)

    def list_suggestions(
        self, filters: SuggestionFilters, *, offset: int, limit: int
    ) -> tuple[list[InternalLinkSuggestion], int]:
        conditions = []
        if filters.status is not None:
            conditions.append(InternalLinkSuggestion.status == filters.status)
        if filters.site_id is not None:
            conditions.append(InternalLinkSuggestion.site_id == filters.site_id)
        if filters.source_page_id is not None:
            conditions.append(InternalLinkSuggestion.source_page_id == filters.source_page_id)
        if filters.target_page_id is not None:
            conditions.append(InternalLinkSuggestion.target_page_id == filters.target_page_id)
        if filters.min_relevance_score is not None:
            conditions.append(InternalLinkSuggestion.relevance_score >= filters.min_relevance_score)
        total = (
            self._session.scalar(
                select(func.count()).select_from(InternalLinkSuggestion).where(*conditions)
            )
            or 0
        )
        items = self._session.scalars(
            select(InternalLinkSuggestion)
            .where(*conditions)
            .order_by(
                InternalLinkSuggestion.created_at.desc(),
                InternalLinkSuggestion.relevance_score.desc(),
                InternalLinkSuggestion.id,
            )
            .offset(offset)
            .limit(limit)
        )
        return list(items), total

    def active_target_ids(self, source_page_id: uuid.UUID) -> set[uuid.UUID]:
        rows = self._session.scalars(
            select(InternalLinkSuggestion.target_page_id).where(
                InternalLinkSuggestion.source_page_id == source_page_id,
                InternalLinkSuggestion.status.in_(ACTIVE_STATUSES),
            )
        )
        return set(rows)

    def rejected_target_ids_since(
        self, source_page_id: uuid.UUID, since: datetime
    ) -> set[uuid.UUID]:
        rows = self._session.scalars(
            select(InternalLinkSuggestion.target_page_id).where(
                InternalLinkSuggestion.source_page_id == source_page_id,
                InternalLinkSuggestion.status == SuggestionStatus.REJECTED,
                InternalLinkSuggestion.updated_at >= since,
            )
        )
        return set(rows)

    def anchors_by_target(self, target_ids: Sequence[uuid.UUID]) -> dict[uuid.UUID, list[str]]:
        if not target_ids:
            return {}
        rows = self._session.execute(
            select(InternalLinkSuggestion.target_page_id, InternalLinkSuggestion.anchor_text).where(
                InternalLinkSuggestion.target_page_id.in_(target_ids),
                InternalLinkSuggestion.status.in_(ACTIVE_STATUSES),
            )
        )
        result: dict[uuid.UUID, list[str]] = {}
        for target_id, anchor in rows:
            result.setdefault(target_id, []).append(anchor)
        return result

    def add_suggestion(self, suggestion: InternalLinkSuggestion) -> bool:
        try:
            with self._session.begin_nested():
                self._session.add(suggestion)
                self._session.flush()
        except IntegrityError as exc:
            # Concurrent analysis already created the active pair (partial unique index).
            if _constraint_name(exc) == ACTIVE_PAIR_INDEX:
                return False
            raise
        return True

    def commit(self) -> None:
        try:
            self._session.commit()
        except IntegrityError as exc:
            self._session.rollback()
            if _constraint_name(exc) == ACTIVE_PAIR_INDEX:
                raise ConflictError(
                    "Another active suggestion already exists for this source and target",
                    code="ACTIVE_SUGGESTION_EXISTS",
                ) from exc
            raise

    def rollback(self) -> None:
        self._session.rollback()

File: src/interlink/repository.js
/** Data access for the interlink module (app/interlink/repository.py). */

import { connect, getPool, runQuery, wrapDbError } from '../db/pool.js';
import { ConflictError, DatabaseError } from '../utils/errors.js';
import { PAGE_COLUMNS } from '../db/queries/pages.js';

export const SuggestionStatus = Object.freeze({
  PENDING: 'PENDING',
  APPROVED: 'APPROVED',
  REJECTED: 'REJECTED',
  APPLIED: 'APPLIED',
});
// A source -> target pair may have at most one suggestion in these states.
export const ACTIVE_STATUSES = [SuggestionStatus.PENDING, SuggestionStatus.APPROVED, SuggestionStatus.APPLIED];
export const ACTIVE_PAIR_INDEX = 'uq_internal_link_suggestions_active_pair';

const SUGGESTION_COLUMNS = `s.id, s.site_id, s.source_page_id, s.target_page_id, s.anchor_text, s.context,
  s.relevance_score, s.reason, s.status::text AS status, s.retrieval_score, s.ai_provider, s.ai_model,
  s.rejection_reason, s.reviewed_at, s.applied_at, s.created_at, s.updated_at`;
const WITH_PAGES = `${SUGGESTION_COLUMNS},
  json_build_object('id', sp.id, 'url', sp.url, 'title', sp.title, 'h1', sp.h1) AS source_page,
  json_build_object('id', tp.id, 'url', tp.url, 'title', tp.title, 'h1', tp.h1) AS target_page`;
const FROM_JOINED = `internal_link_suggestions s
  JOIN pages sp ON sp.id = s.source_page_id
  JOIN pages tp ON tp.id = s.target_page_id`;

/**
 * Unit of work mirroring a SQLAlchemy Session: a dedicated connection, a transaction that begins
 * on first use (autobegin), and explicit `commit()` / `rollback()` that end it.
 */
export class PgSession {
  constructor(pool = getPool()) {
    this.pool = pool;
    this.client = null;
    this.inTransaction = false;
  }

  async query(text, params = []) {
    if (this.client === null) this.client = await connect(this.pool);
    if (!this.inTransaction) {
      await runQuery(this.client, 'BEGIN');
      this.inTransaction = true;
    }
    return runQuery(this.client, text, params);
  }

  async commit() {
    if (this.inTransaction) {
      this.inTransaction = false;
      await runQuery(this.client, 'COMMIT');
    }
  }

  async rollback() {
    if (this.inTransaction) {
      this.inTransaction = false;
      try {
        await this.client.query('ROLLBACK');
      } catch {
        /* connection may be broken */
      }
    }
  }

  async close() {
    await this.rollback();
    if (this.client !== null) {
      this.client.release();
      this.client = null;
    }
  }
}

function isActivePairViolation(err) {
  return err instanceof DatabaseError && err.sqlState === '23505' && err.constraint === ACTIVE_PAIR_INDEX;
}

export class PgInterlinkRepository {
  constructor(session) {
    this.session = session;
  }

  async getPage(pageId, { forUpdate = false } = {}) {
    const { rows } = await this.session.query(
      `SELECT ${PAGE_COLUMNS} FROM pages WHERE id = $1${forUpdate ? ' FOR UPDATE' : ''}`,
      [pageId],
    );
    return rows[0] ?? null;
  }

  /**
   * Cheap SQL pre-filter; `candidate-filter` remains the authoritative rule set. `content_html`
   * is only loaded for pages without title and H1 (needed by the EMPTY_CONTENT rule).
   */
  async listCandidatePool(source) {
    const { rows } = await this.session.query(
      `SELECT id, site_id, url, title, h1, meta_description, content_version, canonical_url, http_status,
              redirect_url, is_indexable, has_noindex, language, region, page_type, keywords, outgoing_links,
              last_crawled_at, created_at, updated_at,
              CASE WHEN btrim(coalesce(title, '')) = '' AND btrim(coalesce(h1, '')) = ''
                   THEN content_html END AS content_html
         FROM pages
        WHERE site_id = $1 AND id <> $2 AND http_status >= 200 AND http_status < 300
          AND redirect_url IS NULL AND has_noindex IS false AND is_indexable IS true
        ORDER BY url`,
      [source.site_id, source.id],
    );
    return rows;
  }

  async loadContents(pageIds) {
    if (!pageIds.length) return new Map();
    const { rows } = await this.session.query('SELECT id, content_html FROM pages WHERE id = ANY($1::uuid[])', [pageIds]);
    const byId = new Map(rows.filter((r) => r.content_html).map((r) => [r.id, r.content_html]));
    // Keep the requested order (Python iterates the DB rows; the order is only used for a dict).
    return new Map(pageIds.filter((id) => byId.has(id)).map((id) => [id, byId.get(id)]));
  }

  async getSuggestion(suggestionId, { forUpdate = false } = {}) {
    const { rows } = await this.session.query(
      `SELECT ${WITH_PAGES} FROM ${FROM_JOINED} WHERE s.id = $1${forUpdate ? ' FOR UPDATE OF s' : ''}`,
      [suggestionId],
    );
    return rows[0] ?? null;
  }

  async listSuggestions(filters, { offset, limit }) {
    const conditions = [];
    const params = [];
    const add = (sql, value) => {
      params.push(value);
      conditions.push(sql.replace('?', `$${params.length}`));
    };
    if (filters.status) add('s.status = ?::interlink_suggestion_status', filters.status);
    if (filters.site_id) add('s.site_id = ?', filters.site_id);
    if (filters.source_page_id) add('s.source_page_id = ?', filters.source_page_id);
    if (filters.target_page_id) add('s.target_page_id = ?', filters.target_page_id);
    if (filters.min_relevance_score !== null && filters.min_relevance_score !== undefined) {
      add('s.relevance_score >= ?', filters.min_relevance_score);
    }
    const where = conditions.length ? `WHERE ${conditions.join(' AND ')}` : '';
    const { rows: countRows } = await this.session.query(
      `SELECT count(*)::int AS total FROM internal_link_suggestions s ${where}`,
      params,
    );
    const { rows } = await this.session.query(
      `SELECT ${WITH_PAGES} FROM ${FROM_JOINED} ${where}
        ORDER BY s.created_at DESC, s.relevance_score DESC, s.id
        OFFSET $${params.length + 1} LIMIT $${params.length + 2}`,
      [...params, offset, limit],
    );
    return [rows, countRows[0].total];
  }

  async activeTargetIds(sourcePageId) {
    const { rows } = await this.session.query(
      `SELECT target_page_id FROM internal_link_suggestions
        WHERE source_page_id = $1 AND status::text = ANY($2::text[])`,
      [sourcePageId, ACTIVE_STATUSES],
    );
    return new Set(rows.map((r) => r.target_page_id));
  }

  async rejectedTargetIdsSince(sourcePageId, since) {
    const { rows } = await this.session.query(
      `SELECT target_page_id FROM internal_link_suggestions
        WHERE source_page_id = $1 AND status = 'REJECTED' AND updated_at >= $2`,
      [sourcePageId, since],
    );
    return new Set(rows.map((r) => r.target_page_id));
  }

  async anchorsByTarget(targetIds) {
    if (!targetIds.length) return new Map();
    const { rows } = await this.session.query(
      `SELECT target_page_id, anchor_text FROM internal_link_suggestions
        WHERE target_page_id = ANY($1::uuid[]) AND status::text = ANY($2::text[])`,
      [targetIds, ACTIVE_STATUSES],
    );
    const result = new Map();
    for (const { target_page_id: id, anchor_text: anchor } of rows) {
      if (!result.has(id)) result.set(id, []);
      result.get(id).push(anchor);
    }
    return result;
  }

  /** Insert; false (without failing the transaction) on an active-pair duplicate. */
  async addSuggestion(s) {
    await this.session.query('SAVEPOINT add_suggestion');
    try {
      const { rows } = await this.session.query(
        `INSERT INTO internal_link_suggestions
           (id, site_id, source_page_id, target_page_id, anchor_text, context, relevance_score, reason, status,
            retrieval_score, ai_provider, ai_model)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9::interlink_suggestion_status, $10, $11, $12)
         RETURNING created_at, updated_at`,
        [
          s.id, s.site_id, s.source_page_id, s.target_page_id, s.anchor_text, s.context, s.relevance_score,
          s.reason, s.status, s.retrieval_score, s.ai_provider, s.ai_model,
        ],
      );
      await this.session.query('RELEASE SAVEPOINT add_suggestion');
      s.created_at = rows[0].created_at;
      s.updated_at = rows[0].updated_at;
      return true;
    } catch (err) {
      const wrapped = wrapDbError(err);
      await this.session.query('ROLLBACK TO SAVEPOINT add_suggestion');
      // Concurrent analysis already created the active pair (partial unique index).
      if (isActivePairViolation(wrapped)) return false;
      throw wrapped;
    }
  }

  /** Persist review/apply field changes (the ORM flush); active-pair clashes -> 409. */
  async saveSuggestion(s) {
    try {
      const { rows } = await this.session.query(
        `UPDATE internal_link_suggestions
            SET status = $2::interlink_suggestion_status, rejection_reason = $3, reviewed_at = $4, applied_at = $5,
                updated_at = now()
          WHERE id = $1
          RETURNING updated_at, reviewed_at, applied_at`,
        [s.id, s.status, s.rejection_reason, s.reviewed_at, s.applied_at],
      );
      Object.assign(s, rows[0]);
    } catch (err) {
      await this.session.rollback();
      if (isActivePairViolation(err)) {
        throw new ConflictError('Another active suggestion already exists for this source and target', {
          code: 'ACTIVE_SUGGESTION_EXISTS',
        });
      }
      throw err;
    }
  }

  async commit() {
    await this.session.commit();
  }

  async rollback() {
    await this.session.rollback();
  }
}
